internal/buildengine/languageplugin/external_plugin_client.go

```diff
@@ -0,0 +1,178 @@
+		if err := p.ping(runCtx); err != nil {
+			cancel()
+			pingErr <- fmt.Errorf("failed to ping plugin")
+		}
+		close(pingErr)
+	}()
+
+	select {
+	case err := <-cmdErr:
+		return err
+	case err := <-pingErr:
+		if err != nil {
+			return nil
+		}
+		return fmt.Errorf("failed to start plugin: %w", err)
+	case <-ctx.Done():
+		return fmt.Errorf("failed to start plugin: %w", ctx.Err())
+	}
+}
+
+func (p *externalPluginImpl) ping(ctx context.Context) error {
+	retry := backoff.Backoff{}
+	heartbeatCtx, cancel := context.WithTimeout(ctx, launchTimeout)
+	defer cancel()
+	err := rpc.Wait(heartbeatCtx, retry, p.client)
+	if err != nil {
+		return connect.NewError(connect.CodeUnavailable, fmt.Errorf("failed to connect to runner: %w", err))
+	}
+	return nil
+}
+
+func (p *externalPluginImpl) kill() error {
+	if err := p.cmd.Kill(syscall.SIGINT); err != nil {
+		return fmt.Errorf("failed to kill language plugin: %w", err)
+	}
+	return nil
+}
+
+func (p *externalPluginImpl) getCreateModuleFlags(ctx context.Context, req *connect.Request[langpb.GetCreateModuleFlagsRequest]) (*connect.Response[langpb.GetCreateModuleFlagsResponse], error) {
+	resp, err := p.client.GetCreateModuleFlags(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get create module flags from plugin: %w", err)
+	}
+	return resp, nil
+}
+
+func (p *externalPluginImpl) moduleConfigDefaults(ctx context.Context, req *connect.Request[langpb.ModuleConfigDefaultsRequest]) (*connect.Response[langpb.ModuleConfigDefaultsResponse], error) {
+	resp, err := p.client.ModuleConfigDefaults(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get module config defaults from plugin: %w", err)
+	}
+	return resp, nil
+}
+
+func (p *externalPluginImpl) createModule(ctx context.Context, req *connect.Request[langpb.CreateModuleRequest]) (*connect.Response[langpb.CreateModuleResponse], error) {
+	resp, err := p.client.CreateModule(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create module: %w", err)
+	}
+	return resp, nil
+}
+
+func (p *externalPluginImpl) getDependencies(ctx context.Context, req *connect.Request[langpb.DependenciesRequest]) (*connect.Response[langpb.DependenciesResponse], error) {
+	resp, err := p.client.GetDependencies(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get dependencies from plugin: %w", err)
+	}
+	return resp, nil
+}
+
+func (p *externalPluginImpl) build(ctx context.Context, req *connect.Request[langpb.BuildRequest]) (chan either.Either[*langpb.BuildEvent, error], streamCancelFunc, error) {
+	stream, err := p.client.Build(ctx, req)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	streamChan := make(chan either.Either[*langpb.BuildEvent, error], 64)
+	go streamToChan(stream, streamChan)
+
+	return streamChan, func() {
+		stream.Close()
+		close(streamChan)
+	}, err
+}
+
+func streamToChan(stream *connect.ServerStreamForClient[langpb.BuildEvent], ch chan either.Either[*langpb.BuildEvent, error]) {
```

Comment:
I don't know of a way to `select` between a stream and a chan in `externalPlugin.run()`, so I convert the stream to a chan here.
LMK if theres a better way

Reply:
That's the best way definitely.